File: dotnet-elasticemail-examples/Examples/Inbound.cs
namespace ElasticEmailExamples;

using ElasticEmail.Api;
using ElasticEmail.Client;
using ElasticEmail.Model;

public static class Inbound
{
    public static async Task RunAsync()
    {
        var inboundApi = new InboundRouteApi(Ee.Config());
        var domain = Ee.SendingDomain;

        // Inbound routing requires the domain's MX record to point at mx.inbound.elasticemail.com.
        // Matching emails are parsed and POSTed as form fields to HttpAddress
        // (from_email, subject, body_text, body_html, att1_name, att1_content, ...).
        // See MinimalApiApp/Program.cs for the receiving handler.

        string? routeId = null;
        try
        {
            var route = await inboundApi.InboundroutePostAsync(new InboundPayload(
                name: "examples-inbound",
                filter: $"*@{domain}",
                filterType: InboundRouteFilterType.EmailAddress,
                actionType: InboundRouteActionType.NotifyViaHttp,
                httpAddress: $"{Ee.PublicUrl}/inbound?token={Uri.EscapeDataString(Ee.WebhookToken)}"));
            routeId = route.PublicId;
            Console.WriteLine($"Inbound route created: {routeId} {route.Filter} -> {route.ActionParameter}");
        }
        catch (ApiException e)
        {
            Ee.Fail("create route", e);
        }

        try
        {
            var routes = await inboundApi.InboundrouteGetAsync();
            Console.WriteLine($"\nInbound routes ({routes.Count}):");
            foreach (var r in routes)
            {
                Console.WriteLine($" - [{r.SortOrder}] {r.PublicId} {r.Name}: {r.FilterType}={r.Filter} {r.ActionType} {r.ActionParameter}");
            }
        }
        catch (ApiException e)
        {
            Ee.Fail("list routes", e);
        }

        // Delete the route we created (comment out to keep it)
        if (routeId != null)
        {
            try
            {
                await inboundApi.InboundrouteByIdDeleteAsync(routeId);
                Console.WriteLine($"\nInbound route deleted: {routeId}");
            }
            catch (ApiException e)
            {
                Ee.Fail("delete route", e);
            }
        }
    }
}

File: elixir-elasticemail-examples/examples/inbound.exs
alias ElasticEmail.Env

public_url = Env.public_url()
token = Env.webhook_token()
domain = Env.sending_domain()

# Inbound routing requires the domain's MX record to point at mx.inbound.elasticemail.com.
# Matching emails are parsed and POSTed as form fields to HttpAddress
# (from_email, subject, body_text, body_html, att1_name, att1_content, ...).
# See phoenix_app/ for the receiving handler.

fail = fn step, error ->
  IO.puts(:stderr, "Error (#{step}): #{ElasticEmail.format_error(error)}")
  System.halt(1)
end

route_id =
  case ElasticEmail.post("/inboundroute", %{
         "Name" => "examples-inbound",
         "Filter" => "*@#{domain}",
         "FilterType" => "EmailAddress",
         "ActionType" => "NotifyViaHttp",
         "HttpAddress" => "#{public_url}/inbound?token=#{URI.encode_www_form(token)}"
       }) do
    {:ok, r} ->
      IO.puts("Inbound route created: #{r["PublicId"]} #{r["Filter"]} -> #{r["ActionParameter"]}")
      r["PublicId"]

    error ->
      fail.("create route", error)
  end

case ElasticEmail.get("/inboundroute") do
  {:ok, routes} ->
    IO.puts("\nInbound routes (#{length(routes)}):")

    Enum.each(routes, fn r ->
      IO.puts(
        " - [#{r["SortOrder"]}] #{r["PublicId"]} #{r["Name"]}: #{r["FilterType"]}=#{r["Filter"]} " <>
          "#{r["ActionType"]} #{r["ActionParameter"] || ""}"
      )
    end)

  error ->
    fail.("list routes", error)
end

# Delete the route we created (comment out to keep it)
if route_id do
  case ElasticEmail.delete("/inboundroute/#{ElasticEmail.path_segment(route_id)}") do
    {:ok, _} -> IO.puts("\nInbound route deleted: #{route_id}")
    error -> fail.("delete route", error)
  end
end
